copycat/api/src/test/java/org/apache/kafka/copycat/data/SchemaProjectorTest.java

```diff
@@ -0,0 +1,464 @@
+                if (optional) {
+                    reader = SchemaBuilder.int8().optional().defaultValue(defaultValue).build();
+                } else {
+                    reader = SchemaBuilder.int8().defaultValue(defaultValue).build();
+                }
+                break;
+            case INT16:
+                if (optional) {
+                    reader = SchemaBuilder.int16().optional().defaultValue(defaultValue).build();
+                } else {
+                    reader = SchemaBuilder.int16().defaultValue(defaultValue).build();
+                }
+                break;
+            case INT32:
+                if (optional) {
+                    reader = SchemaBuilder.int32().optional().defaultValue(defaultValue).build();
+                } else {
+                    reader = SchemaBuilder.int32().defaultValue(defaultValue).build();
+                }
+                break;
+            case INT64:
+                if (optional) {
+                    reader = SchemaBuilder.int64().optional().defaultValue(defaultValue).build();
+                } else {
+                    reader = SchemaBuilder.int64().defaultValue(defaultValue).build();
+                }
+                break;
+            case FLOAT32:
+                if (optional) {
+                    reader = SchemaBuilder.float32().optional().defaultValue(defaultValue).build();
+                } else {
+                    reader = SchemaBuilder.float32().defaultValue(defaultValue).build();
+                }
+                break;
+            case FLOAT64:
+                if (optional) {
+                    reader = SchemaBuilder.float64().optional().defaultValue(defaultValue).build();
+                } else {
+                    reader = SchemaBuilder.float64().defaultValue(defaultValue).build();
+                }
+                break;
+            case BOOLEAN:
+                if (optional) {
+                    reader = SchemaBuilder.bool().optional().defaultValue(defaultValue).build();
+                } else {
+                    reader = SchemaBuilder.bool().defaultValue(defaultValue).build();
+                }
+                break;
+            case BYTES:
+                if (optional) {
+                    reader = SchemaBuilder.bytes().optional().defaultValue(defaultValue).build();
+                } else {
+                    reader = SchemaBuilder.bytes().defaultValue(defaultValue).build();
+                }
+                break;
+            case STRING:
+                if (optional) {
+                    reader = SchemaBuilder.string().optional().defaultValue(defaultValue).build();
+                } else {
+                    reader = SchemaBuilder.string().defaultValue(defaultValue).build();
+                }
+                break;
+            default:
+                break;
+        }
+        Object projected = SchemaProjector.project(writer, reader, value);
+        assertEquals(value, projected);
```

Comment:
With the type promotion changes I think you'll need to pass in both the original value and the projected value to this method and this assertion will have to change to something like `assertEquals(expectedProjected, projected)`.